### PR TITLE
fix(ext/net): return an error from `startTls` and `serveHttp` if the original connection is captured elsewhere 

### LINKED ISSUE #9360: Upgrading Conn to TLS Fails
Hi, I'm trying to upgrade an existing connection to a client socket to use TLS, in this case the SMTP 'STARTTLS' command... There is not much to find about this online, but this is the code I have, and the result bellow.

```ts
logger.trace('Upgrading to STARTTLS.');

await new Reply(220, 'OK, Go ahead!', '2.0.0').send(session.conn);
const tls: Deno.Conn = await Deno.startTls(session.conn);
```

```
Mon, 01 Feb 2021 14:18:10 GMT : (Trace@SMTP/127.0.0.1) » Upgrading to STARTTLS.
thread 'main' panicked at 'Only a single use of this resource should happen: FullDuplexResource { rd: AsyncRefCell<tokio::net::tcp::split_owned::OwnedReadHalf>, wr: AsyncRefCell<tokio::net::tcp::split_owned::OwnedWriteHalf>, cancel_handle: CancelHandle { node: Node { inner: UnsafeCell, _pin: PhantomPinned } } }', runtime/ops/tls.rs:126:6
stack backtrace:
   0: rust_begin_unwind
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/std/src/panicking.rs:495:5
   1: core::panicking::panic_fmt
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/core/src/panicking.rs:92:14
   2: core::option::expect_none_failed
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/core/src/option.rs:1268:5
   3: <futures_util::future::future::Map<Fut,F> as core::future::future::Future>::poll
   4: <futures_util::future::future::Inspect<Fut,F> as core::future::future::Future>::poll
   5: <futures_util::future::future::Map<Fut,F> as core::future::future::Future>::poll
   6: <futures_util::stream::futures_unordered::FuturesUnordered<Fut> as futures_core::stream::Stream>::poll_next
   7: deno_core::runtime::JsRuntime::poll_event_loop
   8: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
   9: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
  10: tokio::runtime::Runtime::block_on
  11: deno::main
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

### LINKED ISSUE #13345: Upgrading connection to TLS panicks
I don't know if closed issues are followed, so I'm creating this issue which relates to #9360 

I'm experiencing the same as that issue with 1.17.2.

It's happening on my Macbook M1. Here's the complete log:
```
thread 'main' panicked at 'Only a single use of this resource should happen: FullDuplexResource { rd: AsyncRefCell<tokio::net::tcp::split_owned::OwnedReadHalf>, wr: AsyncRefCell<tokio::net::tcp::split_owned::OwnedWriteHalf>, cancel_handle: CancelHandle { node: Node { inner: UnsafeCell { .. }, _pin: PhantomPinned } } }', /Users/lucacasonato/projects/github.com/denoland/deno/ext/net/ops_tls.rs:721:6
stack backtrace:
   0:        0x105795e18 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h6db16a647ea8ef75
   1:        0x10509ef74 - core::fmt::write::h33ea04624f802921
   2:        0x105795790 - std::io::Write::write_fmt::hd387c43cb670231f
   3:        0x10579522c - std::panicking::default_hook::{{closure}}::hafe36fc92a4bd1da
   4:        0x1057945e0 - std::panicking::rust_panic_with_hook::hcafd722806bfc4bc
   5:        0x1057ae668 - std::panicking::begin_panic_handler::{{closure}}::h857e2d0c4fb1b7ef
   6:        0x1057ae5e4 - std::sys_common::backtrace::__rust_end_short_backtrace::hef54dc67870da455
   7:        0x1057ae5a0 - _rust_begin_unwind
   8:        0x1066d410c - core::panicking::panic_fmt::h70bba9f1fa3fca77
   9:        0x1066d45d4 - core::result::unwrap_failed::h93f88410b7cbb383
  10:        0x10531e1f4 - <futures_util::future::future::Map<Fut,F> as core::future::future::Future>::poll::hd6aa2ce5e4f45400
  11:        0x10524e128 - <futures_util::future::future::Inspect<Fut,F> as core::future::future::Future>::poll::ha1e027248e82e17e
  12:        0x1050dd2c0 - futures_util::stream::stream::StreamExt::poll_next_unpin::hdad5eebc5c23eb7f
  13:        0x1050d97fc - deno_core::runtime::JsRuntime::poll_event_loop::hb092758295959ea7
  14:        0x104f47638 - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h1fec428dc61032d8
  15:        0x105050be4 - deno::run_command::{{closure}}::h4380190765d113af
  16:        0x104f61358 - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h4fcc6f206a139af5
  17:        0x104f6aea0 - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h6be8160c79fb85ff
  18:        0x1050555b8 - deno::main::h68206ac11b06f611
  19:        0x104e13dc4 - std::sys_common::backtrace::__rust_begin_short_backtrace::h6505599a3d7af1ff
  20:        0x105069dd8 - _main
```

### LINKED ISSUE #13926: ops panic in net/ext with a tls operation
Hi,

Following code of mine caused panic, was instructed to report here. More accuratelly when fetching return value from haeMPYhteenveto function.

```js
const haeMPYhteenveto = () => {
    const mpSumma = itemServices.haeSumma();
    console.log(mpSumma);
    return mpSumma;
};

const haeOstokset = async ({ response }) => {
    const mpSumma = haeMPYhteenveto();
    response.body = await renderFile('../views/kulut.eta', {
        kustannukset: await itemServices.haeHankinnat(),
        mpsumma: mpSumma,
    });
};
```

**Error note:**
```
Deno has panicked. This is a bug in Deno. Please report this
at https://github.com/denoland/deno/issues/new.
If you can reliably reproduce this panic, include the
reproduction steps and re-run with the RUST_BACKTRACE=1 env
var set and include the backtrace in your report.

Platform: windows x86_64
Version: 1.18.1
Args: ["C:\\Users\\Admin\\.deno\\bin\\deno.exe", "run", "--allow-net", "--unstable", "--allow-read", "--allow-write", "app.js"]

thread 'main' panicked at 'Only a single use of this resource should happen: FullDuplexResource { rd: AsyncRefCell<tokio::net::tcp::split_owned::OwnedReadHalf>, wr: AsyncRefCell<tokio::net::tcp::split_owned::OwnedWriteHalf>, cancel_handle: CancelHandle { node: Node { inner: UnsafeCell { .. }, _pin: PhantomPinned } } }', D:\a\deno\deno\ext\net\ops_tls.rs:812:6
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

### LINKED ISSUE #16241: `Deno.serveHttp` always panics under certain conditions
Snippet:

```ts
const port = 12345;

const listener = Deno.listen({ port });

const [serverConn, clientConn] = await Promise.all([
  listener.accept(), 
  Deno.connect({ port }),
]);

const buf = new Uint8Array(128);
const readPromise = serverConn.read(buf);
const http = Deno.serveHttp(serverConn);
```

Running the above code with `deno run --allow-net snippet.ts` always gives us panic.

```
❯ deno run --allow-net servehttp_panic.ts

============================================================
Deno has panicked. This is a bug in Deno. Please report this
at https://github.com/denoland/deno/issues/new.
If you can reliably reproduce this panic, include the
reproduction steps and re-run with the RUST_BACKTRACE=1 env
var set and include the backtrace in your report.

Platform: macos aarch64
Version: 1.26.1
Args: ["deno", "run", "--allow-net", "servehttp_panic.ts"]

thread 'main' panicked at 'Only a single use of this resource should happen: FullDuplexResource { rd: AsyncRefCell<tokio::net::tcp::split_owned::OwnedReadHalf>, wr: AsyncRefCell<tokio::net::tcp::split_owned::OwnedWriteHalf>, cancel_handle: CancelHandle { node: Node { inner: UnsafeCell { .. }, _pin: PhantomPinned } } }', runtime/ops/http.rs:48:8
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

### PR DESCRIPTION
**Pull request by magurotuna**
Fixes #9360
Fixes #13345
Fixes #13926
Fixes #16241

# Overview

This PR removes the calls to `expect()` on `std::rc::Rc`, which caused Deno to panic under certain situations. We now return an error if `Rc` is referenced by other variables.

The point is, our assumption - we have exclusive access to the underlying connection when upgrading it to upper-layer connection - was actually wrong.

# Detail

```ts
const hostname = "localhost";
const port = 25432;

// Start listening on server
const tcpListener = Deno.listen({ hostname, port });

// Establish TCP connection on both server and client
const [serverConn, clientConn] = await Promise.all([
  tcpListener.accept(),
  Deno.connect({ hostname, port }),
]);

const buf = new Uint8Array(128);
// [client] Try to read data from the server (without awaiting)
const readPromise = clientConn.read(buf);
// [client] Upgrade to TLS connection
await Deno.startTls(clientConn, { hostname });
```

Running the above code we can see Deno always panics. This is because `clientConn` is captured by ongoing promise (`readPromise`) although `Deno.startTls` assumes that it has exclusive access to the connection. 

One solution would be we ensure that all of the other promises capturing the connection have been completed before calling `startTls`. This PR basically follows this approach - instead of panicking in such situations, it returns a bad resource error to indicate that you need to check other promises. I'm not sure if the error should be classified as a bad resource though.

This issue is beyond `startTls`. Other builtin functions such as `serveHttp` are also suffering from the issue for the same mechanism.

Looking at this issue a bit further, the root cause is that the operation of upgrading a TCP connection to TLS requires a TCP one to be *consumed* i.e. after upgrading it, the original TCP connection is deleted from the resource table and no longer usable. The requirement for consumption of the original connection is the reason why exclusive access is needed. In fact, this behavior has caused trouble and there's a issue about it:

- #11744

To address this root cause, I think we could clone the original connection (for example using [std::net::TcpStream::try_clone](https://doc.rust-lang.org/std/net/struct.TcpStream.html#method.try_clone)) instead of consuming it, which allows us to retain the original connection in the resource table and to add an upgraded connection as a new resource. In this way, we don't have to consume the original connection and thus the upgrade process won't need exclusive access to the original one.

The problem with this approach is [tokio's TcpStream doesn't have try_clone method](https://github.com/tokio-rs/tokio-core/issues/198). Maybe we could implement the clone logic using a raw file descriptor together with syscall (like `dup`) but it might be platform-specific and need further investigation to see if it works as we expect. So I didn't go with this approach in this PR.

### CODE REVIEW

**Comment by imcotton** [comment]
Thanks for the investigation, in case of issue 11744 mentioned on above, I was hoping that at least the [`readable`][10] and [`writable`][20] from `Conn` works since they're representing modern interface of web streams, thus currently I'm stuck with deprecated `std@0.117.0/http/server_legacy` as workaround, but also worried its removal in upcoming Deno v2.

[10]: https://deno.land/api@v1.26.1?s=Deno.Conn#prop_readable
[20]: https://deno.land/api@v1.26.1?s=Deno.Conn#prop_writable

File: cli/dts/lib.deno.ns.d.ts [hunk 1]
```diff
@@ -4017,6 +4017,14 @@ declare namespace Deno {
    * }
    * ```
    *
+   * Note that this function *consumes* the given connetion passed to it, thus the
```
Comment (bartlomieju) [inline cli/dts/lib.deno.ns.d.ts:4020]:
```suggestion
   * Note that this function *consumes* the given connection passed to it, thus the
```

File: ext/net/lib.deno_net.d.ts [hunk 2]
```diff
@@ -253,9 +253,18 @@ declare namespace Deno {
    * this function requires that the other end of the connection is prepared for
    * a TLS handshake.
    *
+   * Note that this function *consumes* the TCP connetion passed to it, thus the
```
Comment (bartlomieju) [inline ext/net/lib.deno_net.d.ts:256]:
```suggestion
   * Note that this function *consumes* the TCP connection passed to it, thus the
```

File: ext/net/lib.deno_net.d.ts [hunk 3]
```diff
@@ -253,9 +253,18 @@ declare namespace Deno {
...
+   * you need to ensure that the TCP connection is not being used elsewhere when
+   * calling this function in order for the TCP connection to be consumed properly.
+   * For instance, if there is a `Promise` that is waiting for read operation on
+   * the TCP connection to complete, it is considered that the TCP connection is
+   * being used elsewhere. In such a case, this function will fail.
+   *
    * ```ts
    * const conn = await Deno.connect({ port: 80, hostname: "127.0.0.1" });
    * const caCert = await Deno.readTextFile("./certs/my_custom_root_CA.pem");
+   * // `conn` becomes unusable after calling `startTls`
```
Comment (bartlomieju) [inline ext/net/lib.deno_net.d.ts:267]:
```suggestion
   * // `conn` becomes unusable after calling `Deno.startTls`
```

File: core/resources.rs [hunk 4]
```diff
@@ -290,15 +290,23 @@ impl ResourceTable {
   /// Removes a resource of type `T` from the resource table and returns it.
   /// If a resource with the given `rid` exists but its type does not match `T`,
   /// it is not removed from the resource table. Note that the resource's
-  /// `close()` method is *not* called.
+  /// `close()` method is *not* called. Also note that there might be a case where
```
Comment (bartlomieju) [inline core/resources.rs:293]:
```suggestion
  /// `close()` method is *not* called. 
  ///
  /// Also note that there might be a case where
```

File: core/resources.rs [hunk 5]
```diff
@@ -290,15 +290,23 @@ impl ResourceTable {
...
+  /// type `T` from `Rc<T>`.
   pub fn take<T: Resource>(&mut self, rid: ResourceId) -> Result<Rc<T>, Error> {
     let resource = self.get::<T>(rid)?;
     self.index.remove(&rid);
     Ok(resource)
   }
 
   /// Removes a resource from the resource table and returns it. Note that the
-  /// resource's `close()` method is *not* called.
+  /// resource's `close()` method is *not* called. Also note that there might be a
```
Comment (bartlomieju) [inline core/resources.rs:305]:
```suggestion
  /// resource's `close()` method is *not* called. 
  ///
  /// Also note that there might be a
```

File: ext/net/ops_tls.rs [hunk 6]
```diff
@@ -812,12 +812,16 @@ where
...
   let resource_rc = state
     .borrow_mut()
     .resource_table
     .take::<TcpStreamResource>(rid)?;
+  // This TCP connection might be used somewhere else. If it's the case, we cannot proceed with the
+  // process of starting a TLS connection on top of this TCP connection, so we just return a bad
+  // resource error. See also: https://github.com/denoland/deno/pull/16242
   let resource = Rc::try_unwrap(resource_rc)
-    .expect("Only a single use of this resource should happen");
+    .map_err(|_| bad_resource("TCP stream is being used somewhere else"))?;
```
Comment (bartlomieju) [inline ext/net/ops_tls.rs:824]:
```suggestion
    .map_err(|_| bad_resource("TCP stream is currently in use"))?;
```

File: runtime/ops/http.rs [hunk 7]
```diff
@@ -44,8 +45,11 @@ fn op_http_start(
     .resource_table
     .take::<TcpStreamResource>(tcp_stream_rid)
   {
+    // This TCP connection might be used somewhere else. If it's the case, we cannot proceed with the
+    // process of starting a HTTP server on top of this TCP connection, so we just return a bad
+    // resource error. See also: https://github.com/denoland/deno/pull/16242
     let resource = Rc::try_unwrap(resource_rc)
-      .expect("Only a single use of this resource should happen");
+      .map_err(|_| bad_resource("TCP stream is being used somewhere else"))?;
```
Comment (bartlomieju) [inline runtime/ops/http.rs:52]:
```suggestion
      .map_err(|_| bad_resource("TCP stream is currently in use"))?;
```

File: runtime/ops/http.rs [hunk 8]
```diff
@@ -56,8 +60,11 @@ fn op_http_start(
     .resource_table
     .take::<TlsStreamResource>(tcp_stream_rid)
   {
+    // This TLS connection might be used somewhere else. If it's the case, we cannot proceed with the
+    // process of starting a HTTP server on top of this TLS connection, so we just return a bad
+    // resource error. See also: https://github.com/denoland/deno/pull/16242
     let resource = Rc::try_unwrap(resource_rc)
-      .expect("Only a single use of this resource should happen");
+      .map_err(|_| bad_resource("TLS stream is being used somewhere else"))?;
```
Comment (bartlomieju) [inline runtime/ops/http.rs:67]:
```suggestion
      .map_err(|_| bad_resource("TLS stream is currently in use"))?;
```

File: runtime/ops/http.rs [hunk 9]
```diff
@@ -71,8 +78,11 @@ fn op_http_start(
   {
     super::check_unstable(state, "Deno.serveHttp");
 
+    // This UNIX socket might be used somewhere else. If it's the case, we cannot proceed with the
+    // process of starting a HTTP server on top of this UNIX socket, so we just return a bad
+    // resource error. See also: https://github.com/denoland/deno/pull/16242
     let resource = Rc::try_unwrap(resource_rc)
-      .expect("Only a single use of this resource should happen");
+      .map_err(|_| bad_resource("UNIX stream is being used somewhere else"))?;
```
Comment (bartlomieju) [inline runtime/ops/http.rs:85]:
```suggestion
      .map_err(|_| bad_resource("UNIX stream is currently in use"))?;
```

**Review by bartlomieju — APPROVED** [review]
LGTM, good fix!

**Review by victoriaquasar — APPROVED** [review]

**Pull request merged**